compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeParamAnalyzer.java

```diff
@@ -398,6 +405,30 @@ private void findTypeParamInTupleForArray(DiagnosticPos pos, BArrayType expType,
         findTypeParam(pos, expType.eType, tupleElementType, env, resolvedTypes, result);
     }
 
+    private void findTypeParamInUnionForArray(DiagnosticPos pos, BArrayType expType, BUnionType actualType,
```

Comment:
I think we can generalize this and `findTypeParamInUnionForMap`.

```java
    private void findTypeParamInUnion(DiagnosticPos pos, BType expType, BUnionType actualType,
                                      SymbolEnv env, HashSet<BType> resolvedTypes, FindTypeParamResult result) {
        LinkedHashSet<BType> memTypes = new LinkedHashSet<>();
        for (BType type : actualType.getMemberTypes()) {
            switch (type.tag) {
                case TypeTags.ARRAY:
                    memTypes.add(((BArrayType) type).eType);
                    break;
                case TypeTags.MAP:
                    memTypes.add(((BMapType) type).constraint);
                    break;
            }
        }
        BUnionType unionType = BUnionType.create(null, memTypes);
        findTypeParam(pos, expType, unionType, env, resolvedTypes, result);
    }
```

To this we can pass as `expType` the array element type or map constraint type.
e.g.,
```java
findTypeParamInUnion(pos, ((BArrayType) expType).eType, (BUnionType) actualType, env, resolvedTypes, 
                     result);
```

I think this will be the cleaner approach when we're adding checks for the rest of the cases in `findTypeParam` too.

Reply:
I also so some duplicates codes, creating an opportunity for refactoring. Let me think of a way to reduce it. But if we pass `eType` we cannot get member types right ?

```
findTypeParamInUnion(pos, ((BArrayType) expType).eType, (BUnionType) actualType, env, resolvedTypes, 
                     result);
```